common/predictive-text/worker/correction/distance-modeler.ts

```diff
@@ -406,94 +418,116 @@ namespace correction {
+        // It was the final tier - store the node for future reference.
+        this.completedPaths.push(currentNode);
+        
+        // Since we don't modify any other tier, we may simply reinsert the removed tier.
+        this.selectionQueue.enqueue(bestTier);
+
+        return {
+          type: 'complete',
+          cost: currentNode.currentCost,
+          finalNode: currentNode
+        };
+      } else {
+        // Time to construct new edges for the next tier!
+        let nextTier = this.tierOrdering[bestTier.index+1];
+        
+        let inputIndex = nextTier.index;
 
-          let deletionEdges = [];
-          if(!substitutionsOnly) {
-            deletionEdges       = currentNode.buildDeletionEdges(this.inputSequence[inputIndex-1]);
-          }     
-          let substitutionEdges = currentNode.buildSubstitutionEdges(this.inputSequence[inputIndex-1]);
+        let deletionEdges = [];
+        if(!substitutionsOnly) {
+          deletionEdges       = currentNode.buildDeletionEdges(this.inputSequence[inputIndex-1]);
+        }     
+        let substitutionEdges = currentNode.buildSubstitutionEdges(this.inputSequence[inputIndex-1]);
 
-          // Note:  we're live-modifying the tier's cost here!  The priority queue loses its guarantees as a result.
-          nextTier.correctionQueue.enqueueAll(deletionEdges.concat(substitutionEdges));
+        // Note:  we're live-modifying the tier's cost here!  The priority queue loses its guarantees as a result.
+        nextTier.correctionQueue.enqueueAll(deletionEdges.concat(substitutionEdges));
 
-          // So, we simply rebuild the selection queue.
-          this.selectionQueue = new models.PriorityQueue<SearchSpaceTier>(this.QUEUE_SPACE_COMPARATOR, this.tierOrdering);
+        // So, we simply rebuild the selection queue.
+        this.selectionQueue = new models.PriorityQueue<SearchSpaceTier>(this.QUEUE_SPACE_COMPARATOR, this.tierOrdering);
 
-          // We didn't reach an end-node, so we just end the iteration and continue the search.
-        }
+        // We didn't reach an end-node, so we just end the iteration and continue the search.
       }
 
       // If we've somehow fully exhausted all search options, indicate that none remain.
-      return null;
+      return unmatchedResult;
     }
 
     // Current best guesstimate of how compositor will retrieve ideal corrections.
-    *getBestMatches(): Generator<SearchResult[]> { 
+    *getBestMatches(waitMillis?: number): Generator<SearchResult[]> { 
       // might should also include a 'base cost' parameter of sorts?
       let searchSpace = this;
       let currentReturns: {[mapKey: string]: SearchNode} = {};
 
+      // JS measures time by the number of milliseconds since Jan 1, 1970.
+      let timeStart = Date.now();
+      let maxTime: number;
+      if(waitMillis == 0) {
+        maxTime = Infinity;
+      } else if(waitMillis == undefined || Number.isNaN(waitMillis)) { // also covers null.
```

Comment:
I'm a bit confused here about the "also covers null" comment. What is `Number.isNaN(null)` trying to accomplish?

suppose `waitMills = null`:

Try this in a JavaScript console:

```javascript
> null == undefined
true
> Number.isNaN(null)
true
> waitMillis == undefined || Number.isNaN(waitMillis)
true
> null == undefined || Number.isNaN(null)
true
> true || false
true
```

Reply:
```
> Number.isNaN(undefined)
false
> undefined == NaN
false
> null == NaN
false
```

Reply:
Oh, so the `waitMillis == undefined` is what the comment is referring to! I had thought the comment was referring to `Number.isNaN(waitMillis)`. Haha, yeah, I already new about the `null == undefined` trick! https://gist.github.com/eddieantonio/2664cc5902516035949b732bbe19ac2b